Migrate "Orders > Shopping Carts" page

boherm opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/8/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | "Orders > Carts" page migration
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | **Before, please build admin new-theme with `make admin-new-theme` !**<br/>First, active `Carts index` feature flag, and then navigate into Orders -> Carts page.<br/>You can filter data, delete not already ordered carts one by one or by bulk actions.<br/>Multistore mode is supported, and in that case we add Shop column into grid (only for all / group context). 
| Fixed ticket?     | Fixes #10584
| Related PRs       | 
| Sponsor company   | 

**Warning: <br/>For now new Kpi use legacy to retrieve data in all already migrated pages!**

boherm committed:
Add new feature flag to enable/disable new Carts page in bo

boherm committed:
Improve BadgeColumn for Grids

- Add alignement system in column
- Add `badge_type_field` to use column in grid to select badge type directly and no only color (like `color_field` for instance)

boherm committed:
Improve used KPI

- Use DI instead of `Context`
- Add purple color
- Create kpi_row factory for Carts page

kpi

boherm committed:
Update CartController

- Add index action page
- Add solo/bulk delete actions
- Add export into csv

boherm committed:
Add grid and query builder for Carts page

Add cart JS

grid

query

grid

boherm committed:
Add CQRS command to solo/bulk delete carts

zuk3975 commented on src/Core/Grid/Definition/Factory/CartGridDefinitionFactory.php:
```diff
@@ -0,0 +1,345 @@
+     *
+     * @return bool
+     */
+    private function needShopNameColumn(): bool
```
could you move this private method bellow protected ones? (we follow same convention for method scopes as symfony - public goes first, then protected and last private methods)

zuk3975 commented on src/Adapter/Cart/CommandHandler/BulkDeleteCartHandler.php:
```diff
@@ -0,0 +1,67 @@
+        $errors = [];
+        foreach ($command->getCartIds() as $cartId) {
+            try {
+                $cart = $this->getCart($cartId);
```
Could you introduce a CartRepository if it doesn't exist yet, which would extend AbstractObjectModelRepository (take a look at that class, it already has some integrity checks and wrapped in exceptions). THen you can inject the repository in constructor and use $this->cartRepository->get($command->getCartId())

zuk3975 commented on src/Adapter/Cart/CommandHandler/DeleteCartHandler.php:
```diff
@@ -0,0 +1,58 @@
+    public function handle(DeleteCartCommand $command): void
+    {
+        $cartId = $command->getCartId();
+        $cart = $this->getCart($cartId);
```
same as above about, use repository and use
```
$this->cartRepository->get($command->getCartId());
```

zuk3975 commented on src/Adapter/Cart/CommandHandler/DeleteCartHandler.php:
```diff
@@ -0,0 +1,58 @@
+     */
+    public function handle(DeleteCartCommand $command): void
+    {
+        $cartId = $command->getCartId();
```
redundant var
```suggestion
```

zuk3975 commented on src/Adapter/Cart/CommandHandler/DeleteCartHandler.php:
```diff
@@ -0,0 +1,58 @@
+
+        if (!$cart->delete()) {
+            throw new CannotDeleteCartException(sprintf('Failed to delete cart "%s".', $cart->id));
+        }
```
$this->cartRepository->delete($cart). (you can inspire from AbstractObjectModelRepository->deleteObjectModel() method.

zuk3975 commented on src/Core/Grid/Query/CartQueryBuilder.php:
```diff
@@ -0,0 +1,309 @@
+
+            if ('id_order' === $filterName) {
+                $filterValue = strtolower($filterValue);
+                $strNonOrdered = strtolower($this->translator->trans('Non ordered', [], 'Admin.Orderscustomers.Feature'));
+                $strAbandonedCart = strtolower($this->translator->trans('Abandoned cart', [], 'Admin.Orderscustomers.Feature'));
```
I dont really understand what happens here, why you need these translations? how id_order value becomes string? (it should be number)

boherm replied:
In fact, it's aim to reproduce legacy behavior but I'm not a big fan too!
In legacy page, we have only one column for id_order and if cart haven't id_order we display "Not ordered" or "Abandoned cart". We have also a input field to search on this column, and for instance If we type "Not" or "Not ordered" we must display all cart "Not ordered".
Personally, I prefer to add another column for cart status and so provide checkbox list and so we can remove translators things in this class! What do you thing ? :)

boherm replied:
My solution are implemented now in my PR :)

zuk3975 commented on src/PrestaShopBundle/Resources/config/services/adapter/cart.yml:
```diff
@@ -128,3 +128,15 @@ services:
       - name: tactician.handler
         command: 'PrestaShop\PrestaShop\Core\Domain\Cart\Command\AddProductToCartCommand'
+
+  prestashop.adapter.cart.command_handler.delete_cart_handler:
+    class: 'PrestaShop\PrestaShop\Adapter\Cart\CommandHandler\DeleteCartHandler'
```
Not so long ago we started autowiring (but long enough that I must require this change, sorry :smile: ). Could you go through your yml definitons and autowire everything (and use public: false where possilbe).
```suggestion
  PrestaShop\PrestaShop\Adapter\Cart\CommandHandler\DeleteCartHandler:
    autowire: true
    public: false
```

zuk3975 reviewed (changes requested):
Very nice work, I added some comments, that shouldn't take too much time.
Also this PR is missing tests (behats for command handlers and ideally some integration tests for CartController grid action)

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml:
```diff
@@ -87,14 +87,35 @@ services:
+      - '@translator'
+      - '@prestashop.adapter.legacy.kpi_configuration'
+      - '@=service("prestashop.adapter.legacy.context").getAdminLink("AdminCarts", true, {"action": "filterOnlyAbandonedCarts"})'
+      - '@=service("prestashop.adapter.legacy.context").getAdminLink("AdminStats", true, {"ajax": 1, "action": "getKpi", "kpi": "abandoned_cart"})'
```
Please inject the legacy context service here. This expression notation for injection is getting ridiculous. (And for others)

boherm committed:
Fix kpi injections

boherm committed:
Fix delete actions and CartRepository

fix repository

boherm committed:
Move private function in CartGridDefinitionFactory

boherm committed:
Improve carts grid

- Add new column for display & search status
- Fix AbandonedCartKpi to go directly into abandoned cart

fix grid improvement

boherm commented:
I work now on testing ;)

boherm committed:
Add behat tests for deleting carts

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/form/choice_provider.yml:
```diff
@@ -320,3 +320,8 @@ services:
+
+  prestashop.core.form.choice_provider.cart_statuses_choice_provider:
+    class: 'PrestaShop\PrestaShop\Core\Form\ChoiceProvider\CartStatusesChoiceProvider'
+    arguments:
+      - '@translator'
```
autowire

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/grid/accessibility_checker.yml:
```diff
@@ -19,6 +19,9 @@ services:
     class: 'PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\DeleteOrderStatesAccessibilityChecker'
 
+  prestashop.core.grid.action.row.accessibility_checker.delete_cart:
+    class: 'PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\DeleteCartAccessibilityChecker'
```
use fqcn

FabienPapet commented on src/Adapter/Cart/Repository/CartRepository.php:
```diff
@@ -0,0 +1,85 @@
+     * @throws CartException
+     * @throws CannotDeleteOrderedCartException
+     */
+    public function delete(CartId $cartId): void
+    {
+        $cart = $this->get($cartId);
+
+        if ($cart->orderExists()) {
+            throw new CannotDeleteOrderedCartException(sprintf('Cart "%s" with order cannot be deleted.', $cart->id));
+        }
+
+        $this->deleteObjectModel($cart, CannotDeleteCartException::class);
+    }
```
This is custom logic and should not be placed inside the repository. Also, it relies on the legacy, can you update it so it does not rely on the `orderExists` function ?


For example, you can create a findByCartId inside the order repository

FabienPapet commented on src/Adapter/Kpi/AverageOrderValueKpi.php:
```diff
@@ -26,41 +26,73 @@
+     */
+    public function __construct(
+        TranslatorInterface $translator,
+        ConfigurationInterface $configuration,
```
```suggestion
        ShopConfigurationInterface $configuration,
```

FabienPapet commented on src/Adapter/Kpi/AverageOrderValueKpi.php:
```diff
@@ -26,41 +26,73 @@
+            'action' => 'getKpi',
+            'kpi' => 'average_order_value',
+        ]);
+        $helper->refresh = (bool) ($this->configuration->get('AVG_ORDER_VALUE_EXPIRE') < time());
```
```suggestion
        $helper->refresh = (bool) ($this->configuration->getBool('AVG_ORDER_VALUE_EXPIRE') < time());
```

boherm replied:
Hum, but `AVG_ORDER_VALUE_EXPIRE` contain a timestamp in fact and we need to retrieve this data as it is.
Is there something I'm missing?

zuk3975 replied:
I think @FabienPapet missed the comparison `< time`. So in fact it could be configuration->getInt() to compare two integers instead of integer and string (because by default all configuration->get('whatever') will be fetched as string. But I don't block for it, legacy code worked for years comparing that string to int so I dont mind it :smile:

FabienPapet commented on src/Core/Domain/Cart/CartStatusType.php:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Cart status type
+ */
+class CartStatusType
```
```suggestion
class CartStatus
```

FabienPapet commented on src/Core/Grid/Action/Row/AccessibilityChecker/DeleteCartAccessibilityChecker.php:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function isGranted(array $record)
```
```suggestion
    public function isGranted(array $record): bool
```

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml:
```diff
@@ -87,14 +87,33 @@ services:
+
   prestashop.adapter.kpi.abandoned_cart:
     class: 'PrestaShop\PrestaShop\Adapter\Kpi\AbandonedCartKpi'
+    arguments:
+      - '@translator'
+      - '@prestashop.adapter.legacy.kpi_configuration'
+      - '@prestashop.adapter.legacy.context'
+      - '@router'
+      - '@prestashop.core.admin.feature_flag.repository'
 
   prestashop.adapter.kpi.average_order_value:
     class: 'PrestaShop\PrestaShop\Adapter\Kpi\AverageOrderValueKpi'
+    arguments:
+      - '@translator'
+      - '@prestashop.adapter.legacy.kpi_configuration'
+      - '@prestashop.adapter.legacy.context'
 
   prestashop.adapter.kpi.conversion_rate:
     class: 'PrestaShop\PrestaShop\Adapter\Kpi\ConversionRateKpi'
+    arguments:
+      - '@translator'
+      - '@prestashop.adapter.legacy.kpi_configuration'
+      - '@prestashop.adapter.legacy.context'
 
   prestashop.adapter.kpi.net_profit_per_visit:
     class: 'PrestaShop\PrestaShop\Adapter\Kpi\NetProfitPerVisitKpi'
+    arguments:
+      - '@translator'
+      - '@prestashop.adapter.legacy.kpi_configuration'
+      - '@prestashop.adapter.legacy.context'
```
Use autowiring for this

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/domain/cart.yml:
```diff
@@ -0,0 +1,6 @@
+services:
+  _defaults:
+    autowire: true
+    public: true
```
```suggestion
    public: false
```

boherm committed:
Apply suggestions from code review

zuk3975 commented on classes/lang/KeysReference/FeatureFlagLang.php:
```diff
@@ -91,3 +91,7 @@
 trans('Enable / Disable the migrated order states page.', 'Admin.Advparameters.Help');
+
+// Carts index feature flag
+trans('Carts index', 'Admin.Advaparameters.Feature');
```
```suggestion
trans('Carts', 'Admin.Advaparameters.Feature');
```

zuk3975 commented on install-dev/data/xml/feature_flag.xml:
```diff
@@ -24,5 +24,6 @@
     <feature_flag id="tax_rules_group" name="tax_rules_group" label_wording="Tax rule groups" label_domain="Admin.Advparameters.Feature" description_wording="Enable / Disable the migrated tax rules page." description_domain="Admin.Advparameters.Help" state="0" stability="beta" />
     <feature_flag id="customer_threads" name="customer_threads" label_wording="Customer threads" label_domain="Admin.Advparameters.Feature" description_wording="Enable / Disable the migrated customer threads page." description_domain="Admin.Advparameters.Help" state="0" stability="beta" />
     <feature_flag id="order_state" name="order_state" label_wording="Order states" label_domain="Admin.Advparameters.Feature" description_wording="Enable / Disable the migrated order states page." description_domain="Admin.Advparameters.Help" state="0" stability="beta" />
+    <feature_flag id="carts_index" name="carts_index" label_wording="Carts index" label_domain="Admin.Advparameters.Feature" description_wording="Enable / Disable the migrated carts index page." description_domain="Admin.Advparameters.Help" state="0" stability="beta" />
```
```suggestion
    <feature_flag id="carts_index" name="carts_index" label_wording="Carts" label_domain="Admin.Advparameters.Feature" description_wording="Enable / Disable the migrated carts index page." description_domain="Admin.Advparameters.Help" state="0" stability="beta" />
```

zuk3975 commented on classes/lang/KeysReference/FeatureFlagLang.php:
```diff
@@ -91,3 +91,7 @@
+
+// Carts index feature flag
+trans('Carts', 'Admin.Advaparameters.Feature');
+trans('Enable / Disable the migrated carts index page.', 'Admin.Advparameters.Help');
```
Oh, here aswell I think

```suggestion
trans('Enable / Disable the migrated carts page.', 'Admin.Advparameters.Help');
```

boherm replied:
I agree and feature_flag.xml as well!

zuk3975 commented on src/Adapter/Order/Repository/OrderRepository.php:
```diff
@@ -71,4 +73,40 @@ public function get(OrderId $orderId): Order
+    public function findByCartId(CartId $cartId): ?Order
+    {
+        try {
+            $orderId = Db::getInstance()->getValue(
+                'SELECT `id_order` FROM `' . _DB_PREFIX_ . 'orders` WHERE `id_cart` = ' . (int) $cartId->getValue(),
+                false
```
would be cleaner to rely on DBAL/QueryBuilder. smth like (:warning: adjust test the code bellow, its just pseudo code)
```
$this->connection->createQueryBuilder()
    ->select('id_order')
    ->from($this->dbPrefix . 'orders')
    ->where('id_cart = :cartId)
    ->setParameter('cartId', $cartId->getValue)
    ->execute()
    ->fetchOne()
```

zuk3975 commented on src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml:
```diff
@@ -87,14 +87,19 @@ services:
     arguments:
       - "@prestashop.core.localization.locale.context_locale"
+
   prestashop.adapter.kpi.abandoned_cart:
     class: 'PrestaShop\PrestaShop\Adapter\Kpi\AbandonedCartKpi'
```
```suggestion
  PrestaShop\PrestaShop\Adapter\Kpi\AbandonedCartKpi:
```

zuk3975 commented on src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml:
```diff
@@ -87,14 +87,19 @@ services:
+    autowire: true
 
   prestashop.adapter.kpi.average_order_value:
     class: 'PrestaShop\PrestaShop\Adapter\Kpi\AverageOrderValueKpi'
```
```suggestion
  PrestaShop\PrestaShop\Adapter\Kpi\AverageOrderValueKpi:
```

zuk3975 commented on src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml:
```diff
@@ -87,14 +87,19 @@ services:
+    autowire: true
 
   prestashop.adapter.kpi.conversion_rate:
     class: 'PrestaShop\PrestaShop\Adapter\Kpi\ConversionRateKpi'
```
```suggestion
PrestaShop\PrestaShop\Adapter\Kpi\ConversionRateKpi:
```

zuk3975 commented on src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml:
```diff
@@ -87,14 +87,19 @@ services:
     class: 'PrestaShop\PrestaShop\Adapter\Kpi\ConversionRateKpi'
+    autowire: true
 
   prestashop.adapter.kpi.net_profit_per_visit:
     class: 'PrestaShop\PrestaShop\Adapter\Kpi\NetProfitPerVisitKpi'
```
```suggestion
  PrestaShop\PrestaShop\Adapter\Kpi\NetProfitPerVisitKpi:
```

zuk3975 commented on src/PrestaShopBundle/Resources/config/services/adapter/order.yml:
```diff
@@ -285,4 +285,6 @@ services:
 
   prestashop.adapter.order.repository.order_repository:
-    class: PrestaShop\PrestaShop\Adapter\Order\Repository\OrderRepository
+    alias: PrestaShop\PrestaShop\Adapter\Order\Repository\OrderRepository
```
```suggestion
    alias: PrestaShop\PrestaShop\Adapter\Order\Repository\OrderRepository
    deprecated: ~
```

zuk3975 commented on src/PrestaShopBundle/Resources/config/services/core/form/choice_provider.yml:
```diff
@@ -320,3 +320,8 @@ services:
     arguments:
       - '@translator'
+
+  prestashop.core.form.choice_provider.cart_statuses_choice_provider:
+    class: 'PrestaShop\PrestaShop\Core\Form\ChoiceProvider\CartStatusesChoiceProvider'
```
```suggestion
  PrestaShop\PrestaShop\Core\Form\ChoiceProvider\CartStatusesChoiceProvider:
```

zuk3975 commented on src/Core/Domain/Cart/CartStatus.php:
```diff
@@ -0,0 +1,56 @@
+     * Time in seconds representing time before not ordered carts are considered abandoned
+     * (For now: 24h)
+     */
+    public const ABANDONED_CART_EXPIRATION_TIME = 86400;
```
It may be interesting to have this value configurable in BO, but nevermind it would be a new feature I guess. Ok for now :+1:

zuk3975 replied:
Oh, also as this class is only used to store constants (and there is no Create/Edit cart commands where we would use it as VO), it could have a private constructor, so that it is clear that its not intended to be instantiated. Like in this similar purpose class https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Domain/Product/ProductSettings.php#L40

zuk3975 commented on src/Adapter/Order/Repository/OrderRepository.php:
```diff
@@ -71,4 +91,43 @@ public function get(OrderId $orderId): Order
+            }
+
+            return null;
+        } catch (\Exception $e) {
```
use Exception;

```suggestion
        } catch (Exception $e) {
```

zuk3975 commented on src/Core/Form/ChoiceProvider/CartStatusesChoiceProvider.php:
```diff
@@ -0,0 +1,62 @@
+    /**
+     * @return array|void
+     */
+    public function getChoices()
```
I think return type can be added
```suggestion
    public function getChoices(): array
```

zuk3975 commented on src/Core/Grid/Data/Factory/CartGridDataFactory.php:
```diff
@@ -0,0 +1,171 @@
+     *
+     * @return string
+     */
+    private function getOrderLabel(array $record)
```
```suggestion
    private function getOrderLabel(array $record): string
```

zuk3975 commented on src/Core/Grid/Data/Factory/CartGridDataFactory.php:
```diff
@@ -0,0 +1,171 @@
+     *
+     * @return array
+     */
+    private function setRecordData(array $record)
```
```suggestion
    private function setRecordData(array $record): array
```

zuk3975 commented on src/Core/Grid/Data/Factory/CartGridDataFactory.php:
```diff
@@ -0,0 +1,171 @@
+     *
+     * @param array $record
+     *
+     * @return array
```
```suggestion
     * @return array<string, mixed>
```

zuk3975 commented on src/Core/Grid/Data/Factory/CartGridDataFactory.php:
```diff
@@ -0,0 +1,171 @@
+    /**
+     * Modify cart record for grid data.
+     *
+     * @param array $record
```
```suggestion
     * @param array<string, mixed> $record
```

zuk3975 commented on src/Core/Grid/Definition/Factory/CartGridDefinitionFactory.php:
```diff
@@ -0,0 +1,373 @@
+/**
+ * Class CartGridDefinitionFactory builds Grid definition for carts listing.
+ */
+final class CartGridDefinitionFactory extends AbstractFilterableGridDefinitionFactory
```
```suggestion
class CartGridDefinitionFactory extends AbstractFilterableGridDefinitionFactory
```

zuk3975 commented on src/Core/Grid/Query/CartQueryBuilder.php:
```diff
@@ -0,0 +1,293 @@
+     *
+     * @return QueryBuilder
+     */
+    private function getQueryBuilder(array $filters)
```
```suggestion
    private function getQueryBuilder(array $filters): QueryBuilder
```

zuk3975 commented on src/Core/Grid/Query/CartQueryBuilder.php:
```diff
@@ -0,0 +1,293 @@
+     * Apply filters to carts query builder
+     *
+     * @param QueryBuilder $qb
+     * @param array $filters
```
```suggestion
     * @param array<string, mixed> $filters
```

zuk3975 commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/CartController.php:
```diff
@@ -55,14 +60,157 @@
+     *
+     * @return Response
+     */
+    public function indexAction(Request $request, CartFilter $filters)
```
```suggestion
    public function indexAction(Request $request, CartFilter $filters): Response
```

zuk3975 commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/CartController.php:
```diff
@@ -55,14 +60,157 @@
+     *
+     * @return RedirectResponse
+     */
+    public function deleteCartAction(int $cartId)
```
```suggestion
    public function deleteCartAction(int $cartId): RedirectResponse
```

zuk3975 commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/CartController.php:
```diff
@@ -55,14 +60,157 @@
+        return $this->redirectToRoute('admin_carts_index');
+    }
+
+    /**
+     * Provides cart ids from request of bulk action
+     *
+     * @param Request $request
+     *
+     * @return array
+     */
+    private function getBulkCartsFromRequest(Request $request): array
+    {
+        $cartIds = $request->request->get('cart_bulk');
+
+        if (!is_array($cartIds)) {
+            return [];
+        }
+
+        return array_map('intval', $cartIds);
+    }
```
you should soon be able to reuse trait https://github.com/PrestaShop/PrestaShop/pull/32447 to get ids instead of adding new method.

zuk3975 commented on src/PrestaShopBundle/Resources/config/routing/admin/sell/orders/carts.yml:
```diff
@@ -1,3 +1,51 @@
+    _controller: PrestaShopBundle\Controller\Admin\Sell\Order\CartController::exportCartAction
+    _legacy_controller: AdminCarts
+    _legacy_link: AdminCarts:exportcart
+    _legacy_feature_flag: carts_index
```
as you can see you put carts_index on every action (no matter if its index, export or delete..), because once we enable index page, all the actions are enabled (and its ok). So I would say the "_index" is redundant, it should simply be enough having "carts". And we will always only have one flag per page to keep it simple :man_shrugging:

zuk3975 commented on tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php:
```diff
@@ -1151,4 +1157,75 @@ private function getCurrency(string $reference): Currency
+    }
+
+    /**
+     * @Given cart :cartReference should not be deleted
```
```suggestion
     * @Given cart :cartReference should exist
```
this way is applicable to more scenarios without adding additional definitions :man_shrugging:

zuk3975 reviewed (changes requested):
Overall a very decent work. Some "nitpick" requests, but the most important things are the missing return types somewhere and of course resolving conflicts.

boherm commented:
> Overall a very decent work. Some "nitpick" requests, but the most important things are the missing return types somewhere and of course resolving conflicts.

Thanks @zuk3975 for your review! :-)
I will check all your comments as soon as possible!

zuk3975 commented on tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php:
```diff
@@ -1151,4 +1157,75 @@ private function getCurrency(string $reference): Currency
+        $this->cartShouldNotBeDeleted($cartReference);
+    }
+
+    /**
+     * @When I bulk delete following carts:
+     */
+    public function iBulkDeleteFollowingCarts(TableNode $table)
+    {
+        $cartIds = [];
+        foreach ($table->getRows() as $row) {
+            $cartIds[] = (int) SharedStorage::getStorage()->get($row[0]);
+        }
+
+        try {
+            $this->getCommandBus()->handle(new BulkDeleteCartCommand($cartIds));
+        } catch (CartException $e) {
+            $this->setLastException($e);
+        }
+    }
```
arguable (therefore optional) request, but as I mentioned in other PR(https://github.com/PrestaShop/PrestaShop/pull/32302#discussion_r1189933385), it could be simplified to reuse $this->referencesToIds() as the list of deletable entities will probably never be too big to become unreadable.
```suggestion
    /**
     * @When I bulk delete following carts:
     */
    public function iBulkDeleteFollowingCarts(string $cartReferences)
    {
        try {
            $this->getCommandBus()->handle(new BulkDeleteCartCommand($this->referencesToIds($cartReferences)));
        } catch (CartException $e) {
            $this->setLastException($e);
        }
    }
```

zuk3975 commented on tests/Integration/Behaviour/Features/Scenario/Cart/Deleting/delete_cart_from_bo.feature:
```diff
@@ -0,0 +1,59 @@
+      | message             | test                       |
+      | payment module name | dummy_payment              |
+      | status              | Awaiting bank wire payment |
+    When I bulk delete following carts:
+      | dummy_cart1 |
+      | dummy_cart2 |
```
following suggestion above

```suggestion
    When I bulk delete carts "dummy_cart1,dummy_cart2"
```

zuk3975 reviewed (changes requested):
Behats failing. And one comment about making CartStatus class constructor private

zuk3975 commented on src/PrestaShopBundle/Resources/config/services/core/domain/cart.yml:
```diff
@@ -0,0 +1,6 @@
+    autowire: true
+    public: false
+
+  PrestaShop\PrestaShop\Adapter\Cart\Repository\CartRepository:
```
oh this should be defined in adapter/cart.yml as you can see by its namespace (its not in domain). So you can get rid of this file afterwards.

zuk3975 commented on src/PrestaShopBundle/Resources/config/services/core/grid/doctrine_query_builder.yml:
```diff
@@ -530,3 +530,13 @@ services:
+    arguments:
+      - '@prestashop.core.query.doctrine_search_criteria_applicator'
+      - '@prestashop.adapter.shop.context'
+      - "@=service('prestashop.adapter.legacy.context').getContext().shop.id"
+      - "@=service('prestashop.adapter.legacy.context').getContext().shop.getContextShopGroupId()"
```
there is a better way.
Shop ids can be injected into search criteria (by extending `PrestaShop\PrestaShop\Core\Search\ShopFilters` it will have the ShopConstraint inside. There is example how I did it with FeatureQueryBuilder in develop branch already.
But I don't require this change now, Im fine with this for now, and it can be improved later :v:

boherm replied:
Oh, it can be nice to do this improvement now 😄

zuk3975 replied:
as you wish (just the PR might expand with some unexpected changes, but im ok with that :smile:)

zuk3975 reviewed (dismissed):
just this one https://github.com/PrestaShop/PrestaShop/pull/31934/files#r1192088594
(sorry for noticing late) and its ok for me :heavy_check_mark: 

the other one (https://github.com/PrestaShop/PrestaShop/pull/31934/files#r1192094627) could be improved in dedicated PR, not blocking

FabienPapet commented on src/Core/Domain/Cart/CartStatus.php:
```diff
@@ -0,0 +1,63 @@
+/**
+ * Cart status
+ */
+class CartStatus
```
can be an enum

FabienPapet commented on src/Core/Form/ChoiceProvider/CartStatusesChoiceProvider.php:
```diff
@@ -0,0 +1,62 @@
+/**
+ * Class CartStatusesChoiceProvider is responsible for providing cart status choices in cart grid filters.
+ */
+final class CartStatusesChoiceProvider implements FormChoiceProviderInterface
```
I know there is a discussion about FormChoiceProvider being relevant for simple use case like this, I think this one could simply be inside a form type. WDYT ?

ping @jolelievre

boherm committed:
Fix after rebase

MatShir commented:
Might fix this issue https://github.com/PrestaShop/PrestaShop/issues/31385. Check it as a bonus on the QA, not a blocker for QA :)

boherm commented:
Can I close my PR? Stuck since a lot of time.

matks commented:
> Can I close my PR? Stuck since a lot of time.

No, you must keep it open UNTIL THE END OF TIMES!
